Readme.md
# Symfony Storage

[![Latest Version](https://img.shields.io/github/release/php-translation/symfony-storage.svg?style=flat-square)](https://github.com/php-translation/symfony-storage/releases)
[![Total Downloads](https://img.shields.io/packagist/dt/php-translation/symfony-storage.svg?style=flat-square)](https://packagist.org/packages/php-translation/symfony-storage)

This is a PHP Translation storage adapter that uses Symfony's translation writer
and loader.

### Install

```bash
composer require php-translation/symfony-storage
```

### Tests

Static code analysis (fix code style, and test using PHPStan) by running:

```
make static
```

### Documentation

Read the full documentation at [https://php-translation.readthedocs.io](https://php-translation.readthedocs.io/en/latest/).


Changelog.md
# Change Log

The change log describes what is "Added", "Removed", "Changed" or "Fixed" between each release.

## Unreleased

## 2.3.1

### Added

* Allow Symfony 6 by @bocharsky-bw in https://github.com/php-translation/symfony-storage/pull/57

### Fixed

* Update branch alias by @github-actions in https://github.com/php-translation/symfony-storage/pull/56
* CSFixer diff-format is always udiff since 3.0 by @rvanlaak in https://github.com/php-translation/symfony-storage/pull/58
* Update CI PHP and Symfony versions by @rvanlaak in https://github.com/php-translation/symfony-storage/pull/59

## 2.3.0

- Adding support for PHP 8

## 2.2.0

- Adding support for translation files with intl domain suffix

## 2.1.0

- Adding support for `php-translation/common` 3.0

## 2.0.0

- Drop `symfony/translation` < 3.4 support by removing legacy layer
- Add strict type hinting

## 1.1.0

- Remove support of PHP < 7.2
- Remove support of Symfony < 3.4
- Add support for Symfony ^5.0

## 1.0.1

### Fixed

- The `LegacyTranslationReader` should also check for`Translation\SymfonyStorage\TranslationLoader`.

### Changed

- The `Translation\SymfonyStorage\TranslationLoader` was deprecated.

## 1.0.0

### Added

- Support for php-translation/common: 1.0

## 0.5.0

### Added

- Support for php-translation/common: 0.3
- Added the "name" attribute on the "unit" node for XLIFF 2.0.

## 0.4.0

### Added

- Introduced `LegacyTranslationReader` and `LegacyTranslationWriter` to provide BC support for Symfony 2.7 to 3.3.

### Changed

- Travis config.

### Removed

- Removed type annotation for first parameter of `FileStorage::__construct`. Type checks will be done with if-statements
to support legacy code.

## 0.3.3

### Added

- Support for Symfony 4

## 0.3.2

### Added

- Support for Symfony 2.7

## 0.3.1

### Fixed

- Make sure XliffLoader can load from resource

## 0.3.0

### Added

- Added more tests
- Improved loader and dumper of Xliff2.0 meta

### Changed

- We will create an output file if no file exists.
- Xliff2.0 is default format

## 0.2.2

### Added

- Support for options in `XliFfConverter`

## 0.2.1

### Added

- Support for options in `FileStorage`

## 0.2.0

### Added

- `XliffDumper` and `XliffLoader`
- A `XliffConverter` utility class
- The `FileStorage` has support for `TransferableStorage`

## 0.1.1

### Added

- Tests and better validation

## 0.1.0

Init release


src/Dumper/Port/SymfonyPort.php
<?php

/*
 * This file is part of the PHP Translation package.
 *
 * (c) PHP Translation team <tobias.nyholm@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Translation\SymfonyStorage\Dumper\Port;

use Symfony\Component\Translation\MessageCatalogue;

/**
 * This code is moved from the Symfony 3.4 repo. It will be removed when
 * we drop support for Symfony 3.3 and below.
 *
 * @author Tobias Nyholm <tobias.nyholm@gmail.com>
 */
class SymfonyPort
{
    public function dumpXliff2($defaultLocale, MessageCatalogue $messages, $domain, array $options = [])
    {
        $dom = new \DOMDocument('1.0', 'utf-8');
        $dom->formatOutput = true;

        $xliff = $dom->appendChild($dom->createElement('xliff'));
        $xliff->setAttribute('xmlns', 'urn:oasis:names:tc:xliff:document:2.0');
        $xliff->setAttribute('version', '2.0');
        $xliff->setAttribute('srcLang', str_replace('_', '-', $defaultLocale));
        $xliff->setAttribute('trgLang', str_replace('_', '-', $messages->getLocale()));

        $xliffFile = $xliff->appendChild($dom->createElement('file'));
        $xliffFile->setAttribute('id', $domain.'.'.$messages->getLocale());

        foreach ($messages->all($domain) as $source => $target) {
            $translation = $dom->createElement('unit');
            $translation->setAttribute('id', strtr(substr(base64_encode(hash('sha256', $source, true)), 0, 7), '/+', '._'));
            $name = $source;
            if (strlen($source) > 80) {
                $name = substr(md5($source), -7);
            }
            $translation->setAttribute('name', $name);
            $metadata = $messages->getMetadata($source, $domain);

            // Add notes section
            if ($this->hasMetadataArrayInfo('notes', $metadata)) {
                $notesElement = $dom->createElement('notes');
                foreach ($metadata['notes'] as $note) {
                    $n = $dom->createElement('note');
                    $n->appendChild($dom->createTextNode(isset($note['content']) ? $note['content'] : ''));
                    unset($note['content']);

                    foreach ($note as $name => $value) {
                        $n->setAttribute($name, $value);
                    }
                    $notesElement->appendChild($n);
                }
                $translation->appendChild($notesElement);
            }

            $segment = $translation->appendChild($dom->createElement('segment'));

            $s = $segment->appendChild($dom->createElement('source'));
            $s->appendChild($dom->createTextNode($source));

            // Does the target contain characters requiring a CDATA section?
            $text = 1 === preg_match('/[&<>]/', $target) ? $dom->createCDATASection($target) : $dom->createTextNode(
                $target
            );

            $targetElement = $dom->createElement('target');
            if ($this->hasMetadataArrayInfo('target-attributes', $metadata)) {
                foreach ($metadata['target-attributes'] as $name => $value) {
                    $targetElement->setAttribute($name, $value);
                }
            }

            $t = $segment->appendChild($targetElement);
            $t->appendChild($text);

            $xliffFile->appendChild($translation);
        }

        return $dom->saveXML();
    }

    /**
     * @param string     $key
     * @param array|null $metadata
     *
     * @return bool
     */
    private function hasMetadataArrayInfo($key, $metadata = null)
    {
        return null !== $metadata &&
            array_key_exists($key, $metadata) &&
            ($metadata[$key] instanceof \Traversable || is_array($metadata[$key]));
    }
}
